File: samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/QnAMakerBaseDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.AI.QnA;
using Microsoft.Bot.Builder.AI.QnA.Dialogs;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;

namespace Microsoft.BotBuilderSamples.Dialog
{
    /// <summary>
    /// QnAMaker action builder class
    /// </summary>
    public class QnAMakerBaseDialog : QnAMakerDialog
    {
        // Dialog Options parameters
        public const string DefaultCardTitle = "Did you mean:";
        public const string DefaultCardNoMatchText = "None of the above.";
        public const string DefaultCardNoMatchResponse = "Thanks for the feedback.";
        private readonly IBotServices _services;
        private readonly IConfiguration _configuration;
        private readonly string DefaultAnswer = "";

        /// <summary>
        /// Initializes a new instance of the <see cref="QnAMakerBaseDialog"/> class.
        /// Dialog helper to generate dialogs.
        /// </summary>
        /// <param name="services">Bot Services.</param>
        public QnAMakerBaseDialog(IBotServices services, IConfiguration configuration) : base()
        {
            this._configuration = configuration;
            this._services = services;

            if (!string.IsNullOrWhiteSpace(configuration["DefaultAnswer"]))
            {
                this.DefaultAnswer = configuration["DefaultAnswer"];
            }
        }

        protected async override Task<IQnAMakerClient> GetQnAMakerClientAsync(DialogContext dc)
        {
            return _services?.QnAMakerService;
        }

        protected override Task<QnAMakerOptions> GetQnAMakerOptionsAsync(DialogContext dc)
        {
            return Task.FromResult(new QnAMakerOptions
            {
                ScoreThreshold = (float)DefaultThreshold,
                Top = DefaultTopN,
                QnAId = 0,
                RankerType = "Default",
                IsTest = false,

                /*
                * Legacy and V2 preview metadata usage - Uncomment below secrion to apply metadata strictFilters
                */

                /*StrictFilters = new[] { new Metadata { Name= "a", Value ="b" }, new Metadata { Name = "c", Value = "d" }},
                StrictFiltersJoinOperator = JoinOperator.OR,*/

                /*
                * Language Service metadata usage - Uncomment below section to apply filters 
                */

                /*Filters = new Bot.Builder.AI.QnA.Models.Filters
                {
                   MetadataFilter = new Bot.Builder.AI.QnA.Models.MetadataFilter
                   {
                       Metadata = GetMetadata(),
                       LogicalOperation = JoinOperator.AND.ToString()
                   },
                   SourceFilter = (new[] { "GithubSampleActiveLearning.tsv", "SampleActiveLearningImport.tsv" }).ToList(),
                   LogicalOperation = JoinOperator.OR.ToString()
                },*/

                EnablePreciseAnswer = this.IsPreciseAnswerEnabled,

                /*
                * For all v2 and language service bots, IncludeUnstructuredSources is set to true by default
                * To exclude unstructured content from answers, set IncludeUnstructuredSources = false
                */

                IncludeUnstructuredSources = this.IsNotLegacyService
            });
        }

        protected async override Task<QnADialogResponseOptions> GetQnAResponseOptionsAsync(DialogContext dc)
        {
            var defaultAnswerActivity = MessageFactory.Text(this.DefaultAnswer);

            var cardNoMatchResponse = (Activity)MessageFactory.Text(DefaultCardNoMatchResponse);

            var responseOptions = new QnADialogResponseOptions
            {
                ActiveLearningCardTitle = DefaultCardTitle,
                CardNoMatchText = DefaultCardNoMatchText,
                NoAnswer = defaultAnswerActivity,
                CardNoMatchResponse = cardNoMatchResponse,
                DisplayPreciseAnswerOnly = this.DisplayPreciseAnswerOnly
            };

            return responseOptions;
        }

        /// <summary>
        /// Helper method to construct metadata in expected format
        /// </summary>
        /// <returns></returns>
        private System.Collections.Generic.List<System.Collections.Generic.KeyValuePair<string, string>> GetMetadata()
        {
            var metadata = new List<System.Collections.Generic.KeyValuePair<string, string>>();

            // For e.g, metadata pairs "category":"api", "language":"csharp" can be specified as below
            // metadata.Add(new KeyValuePair<string, string>("category", "api"));
            // metadata.Add(new KeyValuePair<string, string>("language", "csharp"));
            return metadata;
        }

        private bool IsPreciseAnswerEnabled
        {
            get
            {
                if (IsNotLegacyService)
                {
                    var rawEnablePreciseAnswer = _configuration["EnablePreciseAnswer"];
                    if (!string.IsNullOrWhiteSpace(rawEnablePreciseAnswer))
                    {
                        return bool.Parse(rawEnablePreciseAnswer);
                    }
                    else
                    {
                        return true;
                    }
                }
                else
                {
                    return false;
                }
            }
        }

        private bool DisplayPreciseAnswerOnly
        {
            get
            {
                var rawDisplayPreciseAnswerOnly = _configuration["DisplayPreciseAnswerOnly"];
                if (!string.IsNullOrWhiteSpace(rawDisplayPreciseAnswerOnly))
                {
                    return bool.Parse(rawDisplayPreciseAnswerOnly);
                }
                else
                {
                    return true;
                }
            }
        }

        private bool IsNotLegacyService
        {
            get
            {
                var qnaServiceType = _configuration["QnAServiceType"];
                if (string.Equals(qnaServiceType, "v2", StringComparison.OrdinalIgnoreCase)
                    || string.Equals(qnaServiceType, "language", StringComparison.OrdinalIgnoreCase))
                {
                    return true;
                }

                return false;
            }
        }
    }
}

File: samples/java_springboot/49.qnamaker-all-features/src/main/java/com/microsoft/bot/sample/qnamaker/all/features/QnAMakerBaseDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.qnamaker.all.features;

import com.microsoft.bot.ai.qna.QnADialogResponseOptions;
import com.microsoft.bot.ai.qna.QnAMakerClient;
import com.microsoft.bot.ai.qna.QnAMakerOptions;
import com.microsoft.bot.ai.qna.dialogs.QnAMakerDialog;
import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.dialogs.DialogContext;
import com.microsoft.bot.integration.Configuration;
import com.microsoft.bot.schema.Activity;
import org.apache.commons.lang3.StringUtils;

import java.util.concurrent.CompletableFuture;

/**
 * QnAMaker action builder class.
 */
public class QnAMakerBaseDialog extends QnAMakerDialog {

    // Dialog Options parameters
    private final String defaultCardTitle = "Did you mean:";
    private final String defaultCardNoMatchText = "None of the above.";
    private final String defaultCardNoMatchResponse = "Thanks for the feedback.";

    private final BotServices services;
    private String defaultAnswer = "No QnAMaker answer found.";

    /**
     * Initializes a new instance of the {@link QnAMakerBaseDialog} class.
     * Dialog helper to generate dialogs.
     *
     * @param withServices Bot Services
     * @param configuration A {@link Configuration} which contains the properties of the application.properties
     */
    public QnAMakerBaseDialog(BotServices withServices, Configuration configuration) {
        super();
        this.services = withServices;
        if (StringUtils.isNotBlank(configuration.getProperty("DefaultAnswer"))) {
            this.defaultAnswer = configuration.getProperty("DefaultAnswer");
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected CompletableFuture<QnAMakerClient> getQnAMakerClient(DialogContext dc) {
        return CompletableFuture.completedFuture(this.services.getQnAMakerService());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected CompletableFuture<QnAMakerOptions> getQnAMakerOptions(DialogContext dc) {
        QnAMakerOptions options = new QnAMakerOptions();
        options.setScoreThreshold(DEFAULT_THRESHOLD);
        options.setTop(DEFAULT_TOP_N);
        options.setQnAId(0);
        options.setRankerType("Default");
        options.setIsTest(false);
        return CompletableFuture.completedFuture(options);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected CompletableFuture<QnADialogResponseOptions> getQnAResponseOptions(DialogContext dc) {
        Activity defaultAnswerActivity = MessageFactory.text(this.defaultAnswer);

        Activity cardNoMatchResponse = MessageFactory.text(this.defaultCardNoMatchResponse);

        QnADialogResponseOptions responseOptions = new QnADialogResponseOptions();
        responseOptions.setActiveLearningCardTitle(this.defaultCardTitle);
        responseOptions.setCardNoMatchText(this.defaultCardNoMatchText);
        responseOptions.setNoAnswer(defaultAnswerActivity);
        responseOptions.setCardNoMatchResponse(cardNoMatchResponse);

        return CompletableFuture.completedFuture(responseOptions);
    }
}
